picard/util/pipe.py

```diff
@@ -0,0 +1,282 @@
+# -*- coding: utf-8 -*-
+#
+# Picard, the next-generation MusicBrainz tagger
+#
+# Copyright (C) 2022 skelly37
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+
+import concurrent.futures
+import os
+from typing import Optional
+
+from picard.const.sys import (
+    IS_MACOS,
+    IS_WIN,
+)
+
+
+if IS_WIN:
+    import win32pipe  # type: ignore
+    import win32file  # type: ignore
+    from pywintypes import error as WinApiError  # type: ignore
+
+
+class PipeError(Exception):
+    pass
+
+
+class PipeErrorInvalidArgs(PipeError):
+    MESSAGE = "ERROR: Pipe() args argument has to be iterable"
+
+    def __init__(self, t):
+        raise PipeError(f"{self.MESSAGE}, {type(t)} provied.")
+
+
+class PipeErrorNotFound(PipeError):
+    MESSAGE = "ERROR: Pipe doesn't exist."
+
+    def __init__(self):
+        raise PipeError(self.MESSAGE)
+
+
+class PipeErrorBroken(PipeError):
+    MESSAGE = "ERROR: Pipe is broken."
+
+    def __init__(self):
+        raise PipeError(self.MESSAGE)
+
+
+class PipeErrorInvalidResponse(PipeError):
+    MESSAGE = "ERROR: Invalid response from pipe:"
+
+    def __init__(self, response):
+        raise PipeError(f"{self.MESSAGE} {response}")
+
+
+class PipeErrorWin(PipeError):
+    MESSAGE = "ERROR: Windows API error\n"
+
+    def __init__(self, winerror):
+        raise PipeError(f"{self.MESSAGE}{winerror}")
+
+
+class Pipe:
+    NO_RESPONSE_MESSAGE: str = "No response from FIFO"
+    MESSAGE_TO_IGNORE: str = "Ignore this message, just testing the pipe"
+    TIMEOUT_SECS: float = 1.5
+
+    def __init__(self, app_name: str, app_version: str, args=None):
+        if args is None:
+            args = tuple()
+        elif args is str:
```

Comment:
If you really want to handle this `str` case (which I advised against in my previous comment), you should use `isinstance()` because one can pass a class inheriting from `str`.

@phw what do you think about this?

Reply:
Imo trying to use `Pipe(n, v, "some_message")` is an intuitive way of trying to send a single message, so I'd like having `isinstance(args, str)` kept.